### ISSUE TITLE
Add support for clipping paths

### ISSUE THREAD

**Issue by keenancrane** [penrose/penrose#376]
A basic way of building up more complex vector graphics from simple primitives is to use clipping paths.  Clipping, for instance, is quite natural in many classes of diagrams—for instance, imagine that you want to plot a collection of lines, curves, etc., but only within a certain rectangular region.

Here's a simple example of how clip paths work in SVG (from [this page](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/clipPath)):
```
<svg viewBox="0 0 100 100">
  <clipPath id="myClip">
    <!--
      Everything outside the circle will be
      clipped and therefore invisible.
    -->
    <circle cx="40" cy="35" r="35" />
  </clipPath>
 
  <!-- The original black heart, for reference -->
  <path id="heart" d="M10,30 A20,20,0,0,1,50,30 A20,20,0,0,1,90,30 Q90,60,50,90 Q10,60,10,30 Z" />
 
  <!--
    Only the portion of the red heart
    inside the clip circle is visible.
  -->
  <use clip-path="url(#myClip)" xlink:href="#heart" fill="red" />
</svg>
```

And here's the resulting image, where the heart is clipped against a circle and drawn in red:

![image](https://user-images.githubusercontent.com/3604525/92046499-ae8f1280-ed50-11ea-8f3d-99c53a3b4692.png)

The question is: how should we expose clipping in Style?  Ideally we'd have a design that supports clipping in its full generality.  E.g., a clipped path can itself be clipped by another path.  I suppose each shape could have an attribute `clipPath`, used like this:

```
x.shape = Circle {
      r: ?
      x: ?
      y: ?
      clipPath: y.shape
}
```

**Comment by keenancrane** [comment]
Can we maybe get this going for just one simple example, e.g., clipping a circle by a circle?

This would help us start "feeling out" what's involved with implementation, without the burden of having to do it globally for all shapes (or deal with SVG groups in full generality).

[Full disclosure: I also actually need circle-circle clipping for a Style I'm writing! :-) ]

**Comment by hypotext** [comment]
Yeah, let's try to get this in before next Weds @wodeni? I found the MDN page pretty confusing; this page is better: https://www.sarasoueidan.com/blog/css-svg-clipping/

To implement it, we need to generate code like this (also from that page):

```
<svg>
    <defs>
        <clipPath id="myClippingPath">
            <!-- ... -->
        </clipPath>
    </defs>
    <!-- the element you want to apply the clipPath to can be any SVG element -->
    <g id="my-graphic" clip-path="url(#myClippingPath)">
        <!-- ... -->
    </g>
</svg>
```

IMO these would be the steps:

- [ ] Add `clipPath: ShapePath` as a property to all shapes in the backend 
- [ ] Extend `Value` with the new type `ShapePath` (it shouldn't literally hold a shape, just be a pointer to one)
- [ ] Add a convention that a shape's name will be its CSS `id` (for referencing/being referenced as a clip path)
- [ ] In each shape's tsx file, have it include its name as its CSS `id`, and also reference its clip paths through `clip-path` attribute
- [ ] In `Canvas.tsx`, generate all the shapes that are used as clipping paths inside one `defs` section for the whole diagram

@wodeni What do you think?

**Comment by wodeni** [comment]
This approach looks reasonable for simple cases (like @keenancrane mentioned). I'm not sure about what kinds of shapes are supported for `<clipPath>` in the SVG spec. I'd suggest double-checking it before you implement it. `<def>` can be a little tricky to get right. You can refer to `Arrow.tsx` for something similar.

**Comment by wodeni** [comment]
Addressed by #1411.
